### PR TITLE
Add Support for Enumerations

### PR DESCRIPTION
**Pull request by vulevukusej**
-Added support for adding and editing custom psets with enumerations, as well as pset templates that have enumerations.

@Moult - Let me know what you think and what needs changing.  After that, I'll continue adding support for the other property types.

### CODE REVIEW

File: src/ifcopenshell-python/ifcopenshell/util/element.py [hunk 1]
```diff
@@ -91,9 +100,9 @@ def get_properties(properties):
 def get_predefined_type(element):
     element_type = get_type(element)
     if element_type:
-        predefined_type = getattr(element_type, "PredefinedType", None)
+        predefined_type = getattr(element_type, "ElementType", getattr(element_type, "ProcessType", None))
         if predefined_type == "USERDEFINED" or not predefined_type:
-            predefined_type = getattr(element_type, "ElementType", getattr(element_type, "ProcessType", None))
+            predefined_type = getattr(element_type, "ElementType", None)
         if predefined_type and predefined_type != "NOTDEFINED":
```
Comment (Moult) [inline src/ifcopenshell-python/ifcopenshell/util/element.py:106]: Is this change correct?

File: src/ifcopenshell-python/ifcopenshell/util/element.py [hunk 2]
```diff
@@ -80,6 +80,15 @@ def get_properties(properties):
     for prop in properties or []:
         if prop.is_a("IfcPropertySingleValue"):
             results[prop.Name] = prop.NominalValue.wrappedValue if prop.NominalValue else None
+        elif prop.is_a("IfcPropertyEnumeratedValue"):
+            values = []
+            for enum in prop.EnumerationValues:
+                values.append(enum.wrappedValue)
```
Comment (Moult) [inline src/ifcopenshell-python/ifcopenshell/util/element.py:86]: values = [v.wrappedValue for v in prop.EnumerationValues]

File: src/ifcopenshell-python/ifcopenshell/util/element.py [hunk 3]
```diff
@@ -80,6 +80,15 @@ def get_properties(properties):
...
+            if values:
+                results[prop.Name] = str(values) if len(values)>1 else values[0]
```
Comment (Moult) [inline src/ifcopenshell-python/ifcopenshell/util/element.py:88]: Why not leave it as a list?

File: src/ifcopenshell-python/ifcopenshell/util/element.py [hunk 4]
```diff
@@ -80,6 +80,15 @@ def get_properties(properties):
...
+            else:
+                results[prop.Name] = None
```
Comment (Moult) [inline src/ifcopenshell-python/ifcopenshell/util/element.py:90]: Code can be condensed to:

results[prop.Name] = [v.wrappedValue for v in prop.EnumerationValues] or None

**Comment by Moult** [comment]
Great job! Only a few minor comments. By the way the PR includes changes for svgfill. That might need to be undone.

File: src/blenderbim/blenderbim/bim/helper.py [hunk 5]
```diff
@@ -33,16 +33,26 @@ def draw_attributes(props, layout, copy_operator=None):
         draw_attribute(attribute, row, copy_operator)
 
 
-def draw_attribute(attribute, layout, copy_operator=None):
+def draw_attribute(box, attribute, layout, copy_operator=None):
```
[the remark this answers is not in the document]
Comment (vulevukusej) [reply src/blenderbim/blenderbim/bim/helper.py:36]: Good spot, that was left behind after messing with some other alternatives.

File: src/ifcopenshell-python/ifcopenshell/util/element.py (hunk 1 above)
Comment (vulevukusej) [reply src/ifcopenshell-python/ifcopenshell/util/element.py:106]: Nope, I must've been working from an older version.

File: src/ifcopenshell-python/ifcopenshell/util/element.py (hunk 3 above)
Comment (vulevukusej) [reply src/ifcopenshell-python/ifcopenshell/util/element.py:88]: Good question - I think I might have (falsely) expected an error from blender if the value wasnt strictly a string type.  I tested with your suggestion and it works fine :)

**Comment by aothms** [comment]
> By the way the PR includes changes for svgfill

Submodules are a pain. But I recommend:

    git config --global submodule.recurse true

That way a `git pull` automatically updates the submodule so that you do not easily revert the changes back by committing the unupdated submodule.

**Comment by vulevukusej** [comment]
@aothms @Moult I assume the best thing to do would be to delete this pull request and make a new one, following the suggestion from @aothms ?

**Comment by aothms** [comment]
Not necessary I think.

If you look at the diff of your pull the relevant change is:

~~~
diff --git a/src/svgfill b/src/svgfill
index f9102b1e9..01a2fb494 160000
--- a/src/svgfill
+++ b/src/svgfill
@@ -1 +1 @@
-Subproject commit f9102b1e9f58e2c3bf9923c1eafa0373eeb872f5
+Subproject commit 01a2fb49434920eb5cce789d14fcb425b9fe82c2
~~~

So what you can do is

~~~
cd src/svgfill
git checkout f9102b1e9f58e2c3bf9923c1eafa0373eeb872f5
cd ..
git add svgfill
git commit -m Submodule
git push
~~~

Go to submodule dir, checkout the previous commit it pointed to, cd up to the main repo, add and commit the submodule and push to the same branch to update the PR.

**Comment by vulevukusej** [comment]
@aothms you're an absolute superstar, thanks 😊.
@Moult Thanks as well, will get started with the other property types now

**Comment by aothms** [comment]
Thanks for taking the time :)

**Pull request merged**